Exporting simulation to CoREAS-style HDF5 files.

galaxyvita opened the pull request:
wrote a function write_CORSIKA7() that is the inverse of the read_CORSIKA7() function. Also wrote all the needed helper functions. Updated the conversion_fieldstrength_cgs_to_SI to use scipy and units only.


Issue being addressed:
Added a function write_CORSIKA7() that allows you to convert .nur file to a hdf5 file.
Added 2 helper functions convert_obs_from_nuradio_efield and convert_obs_positions_from_nuradio_on_ground.
Updated the conversion_fieldstrength_cgs_to_SI to only use scipy and .units.

Commit:
Made multiple functions to convert a .nur file to a hdf5 file

wrote a function write_CORSIKA7() that is the inverse of the
read_CORSIKA7() function. Also wrote all the needed helper functions.
Updated the conversion_fieldstrength_cgs_to_SI to use scipy and units
only.

fschlueter commented on NuRadioReco/modules/io/coreas/coreas.py:
```diff
@@ -21,7 +22,9 @@
-conversion_fieldstrength_cgs_to_SI = 2.99792458e10 * units.micro * units.volt / units.meter
+from scipy.constants import c
+
+conversion_fieldstrength_cgs_to_SI = c*100*units.centimeter/units.second * units.micro * units.volt / units.meter #2.99792458e10 * units.micro * units.volt / units.meter
```
So I think just replacing `2.99792458e10` with `constant.c / units.cm` would work.

fschlueter replied:
Maybe you also want to check this module `NuRadioReco/utilities/constants.py` which defines c in NuRadio units. (Keep in mind that the base unit for time is ns, that means `units.s=1e9`

fschlueter replied:
hm I am a bit confused. So the old variable had a value of `29979.245799999997` which is basically identical with what was used in the original `convert_to_hdf.py` of `conversion_fieldstrength_cgs_to_SI = 2.99792458e4`. This expression amounts to `2.99792458e-07`. Probably because we use the NuRadio units to convert from cgs to SI but for this step I think it might be wrong to use the base unit for time (ns) because we still interpret the voltage in V. @galaxyvita sorry for the hussle but can you maybe just change it back?

galaxyvita replied:
I do not see the mistake?
When I run the following I get the same answer:
`from NuRadioReco.utilities import units
from NuRadioReco.utilities.constants import c

conversion_fieldstrength_cgs_to_SI = c/(units.cm / units.s) * units.micro * units.volt / units.meter
print(conversion_fieldstrength_cgs_to_SI)
>>>29979.245799999993
conversion_fieldstrength_cgs_to_SI = 2.99792458e10 * units.micro * units.volt / units.meter
print(conversion_fieldstrength_cgs_to_SI)
>>>29979.245799999997`

fschlueter replied:
arg sorry, I tested an intermediate version. Ignore my last comment.

Commit:
Updated conversion_fieldstrength_cgs_to_SI because it was wrong.

fschlueter committed:
Update coreas.py

fschlueter commented on NuRadioReco/modules/io/coreas/coreas.py:
```diff
@@ -313,6 +387,105 @@ def read_CORSIKA7(input_file, declination=None, site=None):
+        inputs_grp.attrs["ERANGE"] = np.array([energy_GeV, energy_GeV])
+        inputs_grp.attrs["MAGNET"] = sim_shower.get_parameter(shp.magnetic_field_vector)
+
+        zenith = sim_shower.get_parameter(shp.zenith)
+        azimuth = sim_shower.get_parameter(shp.azimuth)
+        inputs_grp.attrs["PHIP"] = zenith
+        inputs_grp.attrs["THETAP"] = azimuth
```
dont we have to account for the different coordinate system in the azimuth angle?

fschlueter commented on NuRadioReco/modules/io/coreas/coreas.py:
```diff
@@ -313,6 +387,105 @@ def read_CORSIKA7(input_file, declination=None, site=None):
+        coreas_grp.attrs["RotationAngleForMagfieldDeclination"] = sim_shower.get_parameter(shp.magnetic_field_rotation) / units.degree
+
+        B_vec = sim_shower.get_parameter(shp.magnetic_field_vector)
+        coreas_grp.attrs["MagneticFieldVector"] = B_vec
```
that does not exist in the regular files right?

galaxyvita replied:
That was an offline parameter I will just remove it.

fschlueter replied:
ping

fschlueter commented on NuRadioReco/modules/io/coreas/coreas.py:
```diff
@@ -524,7 +696,7 @@ def add_electric_field_to_sim_station(
     electric_field.set_parameter(efp.zenith, zenith)
     electric_field.set_parameter(efp.azimuth, azimuth)
-
+    
```
iiiih

fschlueter commented:
@galaxyvita a few more inline comments. In general it looks very good. One other thing. Do we want to add a module and or script which exports simulations (after the interpolator)?

Commit:
Updated the transformations of the angles and B field because they were still in NRR system

Commit:
updated the magneticfield transformation to take declenation into account

Commit:
Removed some comments

galaxyvita commented on NuRadioReco/modules/io/coreas/coreas.py:
```diff
@@ -314,6 +388,136 @@ def read_CORSIKA7(input_file, declination=None, site=None):
+    with h5py.File(output_file, "x") as f:
+        
+        inputs_grp = f.create_group("inputs")
+        inputs_grp.attrs["RUNNR"] = evt.get_run_number()
```
Added

Commit:
Updated some small changed

removed MagneticFieldVector, added EVTNR, changed the function name to
convert_nuradio_efield_to_obs.

fschlueter reviewed (dismissed):
I am assuming you tested this function? E.g., exported a (interpolated) simulation and read it in again?

Commit:
Changed convert_obs_positions_from_nuradio_on_ground name to convert_nuradio_positions_to_ground

galaxyvita commented on NuRadioReco/modules/io/coreas/coreas.py:
```diff
@@ -219,6 +261,38 @@ def convert_obs_positions_to_nuradio_on_ground(observer_pos, declination=0):
 
     return np.squeeze(obs_positions)
 
+def convert_obs_positions_from_nuradio_on_ground(geo_pos, declination=0):
```
Done

Commit:
type in the angles

galaxyvita commented:
The traces do look the same but the values aren't exactly the same which I think might be because of the angles not being the same. Doing from 60° to radians and then back to degrees gives you and angle of 59°.

galaxyvita commented:
Yes I deleted that comments because I realised after that what I said was wrong. But the comment was that I wrote a function to check if the positions and fields are the same but I didn't put the tolerance on the Efield high enough to notice the difference.
The function was this:
`def match_and_compare_observers(file1, file2, tol_pos=1e-6, tol_field=1e-11):

    with h5py.File(file1, "r") as f1, h5py.File(file2, "r") as f2:
        obs_grp_1 = f1["CoREAS/observers"]
        obs_grp_2 = f2["CoREAS/observers"]

        positions_1 = []
        for name in obs_grp_1:
            pos = np.array(obs_grp_1[name].attrs["position"])
            positions_1.append((name, pos))

        positions_2 = []
        for name in obs_grp_2:
            pos = np.array(obs_grp_2[name].attrs["position"])
            positions_2.append((name, pos))

        matched = []
        unmatched_1 = set(name for name, _ in positions_1)
        unmatched_2 = set(name for name, _ in positions_2)

        for name1, pos1 in positions_1:
            for name2, pos2 in positions_2:
                if np.allclose(pos1, pos2, atol=tol_pos):
                    matched.append((name1, name2))
                    unmatched_1.discard(name1)
                    unmatched_2.discard(name2)
                    break

        print(f"Matched {len(matched)} observers:")
        for n1, n2 in matched:
            efield1 = np.array(obs_grp_1[n1])
            efield2 = np.array(obs_grp_2[n2])

            same_pos = np.allclose(obs_grp_1[n1].attrs["position"],
                                   obs_grp_2[n2].attrs["position"], atol=tol_pos)
            same_field = np.allclose(efield1, efield2, atol=tol_field)

            print(f"  {n1} ↔ {n2}")
            print(f"    Position match: {same_pos}")
            print(f"    E-field match:  {same_field}")

        if unmatched_1:
            print(f"\nObservers only in {file1}: {unmatched_1}")
        if unmatched_2:
            print(f"Observers only in {file2}: {unmatched_2}")`

galaxyvita commented:
Thank you for the information. I do not think I have the time right now to learn how to do that so it would be very much appreciated if you could write that. 
I also wrote a function that compares the attributes of the 2 hdf5 files, of course this will print a lot of stuff because you aren't saving everything.

`def compare_hdf5(file1, file2, tol=1e-9):
    def compare_groups(g1, g2, path="/"):

        if path.startswith("/CoREAS/observers/"):
            return

        for k, v in g1.attrs.items():
            if k not in g2.attrs:
                print(f"Missing attribute '{k}' in {path} of {file2}")
                continue
            if isinstance(v, np.ndarray):
                if not np.allclose(v, g2.attrs[k], rtol=tol, atol=tol):
                    print(f"Attr mismatch in {path}/{k}")
            else:
                if v != g2.attrs[k]:
                    print(f"Attr mismatch in {path}/{k}: {v} != {g2.attrs[k]}")

        for k in g2.attrs:
            if k not in g1.attrs:
                print(f"Extra attribute '{k}' in {path} of {file2}")

        names1 = set(g1.keys())
        names2 = set(g2.keys())

        for name in names1 - names2:
            print(f"Missing dataset/group {path}{name} in {file2}")
        for name in names2 - names1:
            print(f"Extra dataset/group {path}{name} in {file2}")

        for name in sorted(names1 & names2):
            item1 = g1[name]
            item2 = g2[name]
            if isinstance(item1, h5py.Dataset) and isinstance(item2, h5py.Dataset):
                arr1 = item1[()]
                arr2 = item2[()]
                if arr1.shape != arr2.shape:
                    print(f"Shape mismatch at {path}{name}: {arr1.shape} != {arr2.shape}")
                elif not np.allclose(arr1, arr2, rtol=tol, atol=tol):
                    print(f"Dataset mismatch at {path}{name}")
            elif isinstance(item1, h5py.Group) and isinstance(item2, h5py.Group):
                compare_groups(item1, item2, path + name + "/")
            else:
                print(f"Type mismatch at {path}{name}")

    with h5py.File(file1, "r") as f1, h5py.File(file2, "r") as f2:
        compare_groups(f1, f2)`

Commit:
Typo in my angles

Commit:
Turned the angles into an array just like in the CoREAS HDF5

Commit:
Removed pint statments

fschlueter commented:
@galaxyvita ?

Commit:
Added the write_CORSIKA7 to the example

fschlueter committed:
Merge branch 'develop' into feature/write-nur-to-hdf5

fschlueter commented:
Or can you write a minimal script which just generates an event using the cr interpolator? I do think having a script which only generates simulated pulses from the interpolater and exports them to hdf5 is acutally a very-nice-to-have for air showers!

galaxyvita commented:
When I run `python sim_cr_det_array.py` I get the following error.
error:
`Traceback (most recent call last):
  File "/ada_mnt/ada/user/vitaldehenau/Github/NuRadioMC/NuRadioReco/examples/cr_analysis/sim_cr_det_array.py", line 108, in <module>
    efieldToVoltageConverter.run(evt, sta, det)
  File "/ada_mnt/ada/user/vitaldehenau/Github/NuRadioMC/NuRadioReco/modules/base/module.py", line 103, in register_run_method
    res = run(self, *args, **kwargs)
  File "/ada_mnt/ada/user/vitaldehenau/Github/NuRadioMC/NuRadioReco/modules/efieldToVoltageConverter.py", line 151, in run
    max_channel_trace_length = np.max([
  File "/ada_mnt/ada/user/vitaldehenau/Github/NuRadioMC/NuRadioReco/modules/efieldToVoltageConverter.py", line 152, in <listcomp>
    det.get_number_of_samples(station.get_id(), channel_id) / det.get_sampling_frequency(station.get_id(), channel_id)
  File "/ada_mnt/ada/user/vitaldehenau/Github/NuRadioMC/NuRadioReco/detector/detector_base.py", line 912, in get_number_of_samples
    return res['adc_n_samples']
KeyError: 'adc_n_samples'`

So I think something changed and now that script is outdated.

Commit:
conversion_fieldstrength_cgs_to_SI fully into NuRadio units

fschlueter commented:
Looks like the last commit was not intended for this branch

galaxyvita commented:
I pushed it here because I used this PR to work in the coreas file and back in the past I also updated the constant here.

Commit:
Reverted all the things i deleted somehow.

Commit:
Updated code based on felix comment.

fschlueter committed:
Merge branch 'develop' into feature/write-nur-to-hdf5

fschlueter commented on NuRadioReco/examples/cr_analysis/sim_cr_det_array.py:
```diff
@@ -128,4 +130,13 @@
+    for idx, evt in enumerate(readCoREASDetector.run(det, NuRadioReco.modules.io.coreas.readCoREASDetector.get_random_core_positions(-300, 0, 1600, 1800, 3))):
+        run_number = evt.get_run_number() if hasattr(evt, "get_run_number") else idx
+        corsika_output = f"SIM{run_number:06d}_{idx}.hdf5"
+        NuRadioReco.modules.io.coreas.coreas.write_CORSIKA7(evt, corsika_output)
```
wont that create new events (and redo the interpolation work)?

fschlueter committed:
Enhance export to CORSIKA with pulse interpolator

Updated export functionality to include pulse interpolator and removed redundant export loop.

Commit:
Wrote an example that uses my new function

Commit:
added the TimeResolution

The code makes sure it is the same TimeResolution in every trace.

Commit:
typo

fschlueter committed:
white space

sjoerd-bouma committed:
attempt to fix cr reading example

sjoerd-bouma committed:
fix error messages raised by RNO-G hardwareResponseIncorporator

sjoerd-bouma committed:
very minor tidying up in coreas.py

sjoerd-bouma committed:
[DOC] various small documentation/rst fixes

Commit:
Added a working example for SKA-Low and now the new hdf5 saves absolute positions.

fschlueter commented on NuRadioReco/examples/Interpolated_SKA_Low_event_to_hdf5.py:
```diff
@@ -1,93 +1,103 @@
-import logging
 
 import NuRadioReco.modules.io.coreas.readCoREASDetector
+import NuRadioReco.modules.io.coreas.readCoREASShower
```
can be deleted?

galaxyvita replied:
a yes.

fschlueter commented on NuRadioReco/modules/io/coreas/coreas.py:
```diff
@@ -438,55 +429,41 @@ def write_CORSIKA7(evt, output_file, declination=None, site=None):
-            By_corsika    = np.dot(magnetic_field_vector_NNR, y_mag) / (units.micro * units.tesla) # along magnetic north
-            minBz_corsika = -np.dot(magnetic_field_vector_NNR, z_hat) / (units.micro * units.tesla) # down component
+            By_corsika = np.dot(b_vec_nnr, y_mag) / (units.micro * units.tesla)
+            minBz_corsika = -np.dot(b_vec_nnr, z_hat) / (units.micro * units.tesla)
```
typo, or change `B_vec_NNR = sim_shower.get_parameter(shp.magnetic_field_vector)`

galaxyvita committed:
Update Interpolated_SKA_Low_event_to_hdf5.py

Removed unneeded import

galaxyvita committed:
Update coreas.py

fschlueter commented on NuRadioReco/examples/Interpolated_SKA_Low_event_to_hdf5.py:
```diff
@@ -0,0 +1,102 @@
+import numpy as np
```
Can you add a short description what this script does?

fschlueter commented on NuRadioReco/examples/cr_analysis/sim_cr_det_array.py:
```diff
@@ -43,89 +24,123 @@
+        if args.export_corsika:
+            run_number = evt.get_run_number() if hasattr(evt, "get_run_number") else idx
+            corsika_output_file = f"SIM{run_number:06d}_{idx}.hdf5"
+            NuRadioReco.modules.io.coreas.coreas.write_CORSIKA7(evt, corsika_output_file)
```
Do we need to add the detector here? Shall we remove the export from this script again since we added a dedicated script?

fschlueter commented on NuRadioReco/modules/io/coreas/coreas.py:
```diff
@@ -314,9 +390,145 @@ def read_CORSIKA7(input_file, declination=None, site=None):
+            else:
+                try:
+                    station_abs_pos = station.get_position() if hasattr(station, 'get_position') else np.zeros(3)
+                except Exception:
```
you can remove that. Only keep `station_abs_pos = np.zeros(3)` if detector is None

fschlueter commented on NuRadioReco/modules/io/coreas/coreas.py:
```diff
@@ -314,9 +390,145 @@ def read_CORSIKA7(input_file, declination=None, site=None):
+                    station_abs_pos = np.zeros(3)
+
+            if station_abs_pos is None:
+                station_abs_pos = np.zeros(3)
```
that should not be necessary

fschlueter reviewed (commented):
Awesome work @galaxyvita, its almost ready!

galaxyvita committed:
Update Interpolated_SKA_Low_event_to_hdf5.py

added a doc string

galaxyvita committed:
Update sim_cr_det_array.py

removed the convert back to hdf5 part because we made a new example for this.

galaxyvita committed:
Update coreas.py

Turned out that those lines didn’t do anything

galaxyvita committed:
Update Interpolated_SKA_Low_event_to_hdf5.py

Cleared the doc string a bit more.

fschlueter commented on NuRadioReco/examples/Interpolated_SKA_Low_event_to_hdf5.py:
```diff
@@ -1,9 +1,26 @@
 """
 An example showing how to use the new write_CORSIKA7() function which stores an event as a hdf5 file.
+This function is intended to make the sharing of realistic simulations (interpolated to a detector, noise added, triggers added, ...) easier.
```
I do not think we want to add noise or simulate a trigger for a to HDF5 exported file?

galaxyvita replied:
Depends if it is easy to add noise afterwards. But the intent of the function could indeed be clarified in the docstring.

galaxyvita committed:
Update Interpolated_SKA_Low_event_to_hdf5.py

fschlueter committed:
Remove non existing interface

fschlueter commented on NuRadioReco/modules/io/coreas/coreas.py:
```diff
@@ -314,9 +390,139 @@ def read_CORSIKA7(input_file, declination=None, site=None):
+        Name of the site, used to set default magnetic field and declination.
+    detector : NuRadioReco.detector.detector.Detector, optional
+        Detector object used to accurately retrieve absolute station positions.
+    """
```
Hey @galaxyvita sorry me again. I was thinking, do we need to pass declination and site here? Should we be  able to use the magnetic field vector stored in the shower parameter storage and infer the declination from it?

I understand if you do not want to work anymore on it but would be happy if you tell me if that makes sense (I can do the necessary changes)

galaxyvita replied:
That would actually be clear if someone does a simulation at a ‘random’ location.
I think i fixed it, the coordinate system was not clear to me but pretty sure it should be the NNR one.

fschlueter commented on NuRadioReco/examples/Interpolated_SKA_Low_event_to_hdf5.py:
```diff
@@ -0,0 +1,125 @@
+        # Both Reconstructed HDF5 and Original Star-Shape HDF5 
+        # already contain absolute ground coordinates in the e-field.
+        if is_reconstructed:
+            abs_pos = np.zeros(3)
```
is this still correct?

fschlueter commented on NuRadioReco/examples/Interpolated_SKA_Low_event_to_hdf5.py:
```diff
@@ -0,0 +1,125 @@
+
+    # Read back and extract reconstructed fluence data
+    reco_evt = coreas.read_CORSIKA7(reconstructed_HDF5)
+    x_reco, y_reco, f_reco, id_reco = get_fluence_data(reco_evt, is_reconstructed=True)
```
can we change the name/prefix `reconstructed` - that is misleading?

Commit:
simplifief based an felix comments

Commit:
calc the declenation from the magnetic-field directly

fschlueter committed:
Refactor station position retrieval logic

Removed redundant exception handling for station position retrieval.
